docs/rfcs/PipelineComposition.md

```diff
@@ -0,0 +1,170 @@
+# RFC: High Level MLIR Pass Pipeline Composition
+
+Ivan Butygin, Renato Golin
+
+## Summary
+
+We propose a new API to build a dependencies-based graph of pass pipelines.
+
+## Motivation
+
+At this point MLIR pass infrastruture only allows to define a linear sequence of passes. While this approach works
+for simple cases, more complex compiler pipelines may require more control over passes execution order.
+
+Two main usecases we are considering:
+* Extensibility. We want to have the abilty for users to extend existing pipeline by inserting theirs custom passes
+into various places while reusing most of the exiting pipeline. With current approach the most common way way to achieve this is
+for pipeline developer to add a fixed 'extension point' during initial pipeline design.
+* Dynamic pipeline control flow. It's often required to have an ability select specific sequence of passes based on some info which
+is only becomes available durin compilation process. It can be either branching to some separate sequnces of passes based on selected codegen path
+or running sequence of passes in the loop until some runtime condition is reached.
+
+## Proposal
+
+This API also allows a dynamic control flow between pipelines in graph, controlled by passes inside said pipelines.
+
+## New APIs
+
+### PipelineGraph
+```C++
+class PipelineGraph {
+public:
+    void registerPipeline(
+        StringRef name,
+        ArrayRef<StringRef> predecessors,
+        ArrayRef<StringRef> successors,
```

Comment:
I'd also avoid having both `predecessors` and `sucessors`. This feels like a duplication and hard to get right on larger graphs.

What I had in mind is just:
* Dependencies: Passes that you _must_ run before (analyses and transforms)
* Post-clean up: Canonicalization that can help the following passes

Dependencies can be bundles or specific passes. Bundles can be just a list of passes (ex. buff+ownership), a loop or a conditional (see below). Both bundles and passes have deps/cleanups and we can simplify the graph after linearization.

Post-cleanups would also be simplified (de-duped) if one pass lists it as its cleanups and the following pass lists it as its dependencies.

Reply:
Regarding having both predecessors and successors,

Following (hypothetical) pipeline:
```
numpy-to-linalg  torch-to-linalg
            \      /
          bufferization
            /      \
  linalg-to-cpu  linalg-to-gpu
```
We don't want to bufferization to know about specific `***-to-linalg` pipelines, as it is a frontend details, irrelevant to bufferization, and we don't want it to know about `linalg-to-***` either as it backend details.
So pipeline should looks like
```
numpy-to-linalg: [], [bufferization]
torch-to-linalg: [], [bufferization]
bufferization: [], []
linalg-to-cpu: [bufferization], []
linalg-to-gpu: [bufferization], []
```